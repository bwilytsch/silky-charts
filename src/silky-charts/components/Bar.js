import React, { useState, useRef, useEffect, useMemo } from 'react';
import identity from 'ramda/src/identity';
import { select as d3Select } from 'd3-selection';
import { max as d3Max } from 'd3-array';
import { axisBottom as d3AxisBottom, axisLeft as d3AxisLeft } from 'd3-axis';
import { Axis, BarDatum, Grid, SVG } from './styled';
import {
  allDate,
  drawGrid,
  extendXPath,
  getBaseColor,
  getDivergence,
  getId,
  getSize,
  getXScale,
  getYScale,
  rotateXLabels,
  setupData,
  valueFor,
} from '../utils';
import {
  SCALE_TIME,
  SCALE_BAND,
  SCALE_LINEAR,
  THEME,
  TICKS,
  MARGIN,
  ASPECT_RATIO,
  SIZE,
} from '../utils/constants';

const Bar = ({
  aspectRatio = ASPECT_RATIO,
  data: chartData,
  showValue,
  showDivergence,
  grid,
  horizontal,
  margin = MARGIN,
  onClick = identity,
  onMouseEnter = identity,
  onMouseLeave = identity,
  width: svgWidth = undefined,
  height: svgHeight = undefined,
  theme = THEME,
  ticks = TICKS,
  xAxisLabelRotation,
  xAxisLabelRotationValue = -50,
}) => {
  const svgRef = useRef();
  const [{ width, height, isSizeSet }, setSize] = useState(SIZE);
  const [currentValue, setCurrentValue] = useState(null);
  const [isDates, data] = useMemo(() => setupData(chartData), chartData);
  const [id] = useState(getId('bar'));

  const xScale = getXScale(
    isDates ? SCALE_TIME : SCALE_BAND,
    data,
    width,
    true
  );
  const yScale = getYScale(
    SCALE_LINEAR,
    d3Max(data, ({ value }) => value),
    height
  );

  const handleOnMouseEnter = event => {
    setCurrentValue(event.target.getAttribute('value'));
    onMouseEnter(event);
  };

  const handleOnMouseLeave = event => {
    setCurrentValue(null);
    onMouseLeave(event);
  };

  const handleResize = () => {
    const offsetWidth = svgRef.current.parentElement.offsetWidth;
    if ((svgWidth || svgHeight) && !isSizeSet) {
      setSize({
        ...getSize(svgWidth, svgHeight, margin, aspectRatio),
        isSizeSet: true,
      });
    } else if (offsetWidth !== svgWidth - (margin.left + margin.right)) {
      setSize({
        ...getSize(offsetWidth, undefined, margin, aspectRatio),
        isSizeSet: true,
      });
    }
  };

  useEffect(() => {
    handleResize();
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <SVG
      ref={svgRef}
      identifier={id}
      size={{
        width: svgWidth || width + margin.left + margin.right,
        height: svgHeight || height + margin.top + margin.bottom,
      }}
    >
      <g
        className="container"
        transform={`translate(${margin.left}, ${margin.top})`}
      >
        {grid && (
          <Grid
            ref={node =>
              d3Select(node).call(
                drawGrid(horizontal, xScale, height, yScale, width, ticks)
              )
            }
          />
        )}

        <Axis
          axis="x"
          position={{ x: 0, y: height }}
          ref={node => {
            d3Select(node).call(d3AxisBottom(xScale));
            isDates && extendXPath(id, width);
            xAxisLabelRotation && rotateXLabels(id, xAxisLabelRotationValue);
          }}
        />
        <Axis
          axis="y"
          ref={node => d3Select(node).call(d3AxisLeft(yScale).ticks(ticks))}
        />

        <g className="data">
          {data.map(({ name, value }, idx) => (
            <BarDatum
              key={idx}
              showValue={showValue}
              datum={{
                name,
                value:
                  showValue && showDivergence && currentValue
                    ? getDivergence(value, currentValue)
                    : value,
              }}
              x={
                isDates
                  ? xScale(name) - valueFor('x', width, data.length)
                  : xScale(name)
              }
              y={yScale(value)}
              width={
                isDates
                  ? valueFor('width', width, data.length)
                  : xScale.bandwidth()
              }
              height={height - yScale(value)}
              color={getBaseColor(theme)}
              onClick={onClick}
              onMouseEnter={handleOnMouseEnter}
              onMouseLeave={handleOnMouseLeave}
            />
          ))}
        </g>
      </g>
    </SVG>
  );
};

export default Bar;
