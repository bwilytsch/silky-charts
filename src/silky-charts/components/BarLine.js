import React, { useEffect, useMemo, useRef, useState } from 'react';
import head from 'ramda/src/head';
import identity from 'ramda/src/identity';
import { axisBottom as d3AxisBottom, axisLeft as d3AxisLeft } from 'd3-axis';
import { select as d3Select } from 'd3-selection';
import { line as d3Line } from 'd3-shape';
import { Axis, Grid, Label, LineDatum, StackedBarDatum, SVG } from './styled';
import {
  buildStack,
  debounce,
  drawGrid,
  extendXPath,
  getId,
  getLineDataForSeries,
  getMax,
  getStackedMax,
  getSize,
  getXScale,
  getYScale,
  palette,
  rotateXLabels,
  setLineType,
  setupData,
  toStackedForm,
  toSingleForm,
} from '../utils';
import {
  ASPECT_RATIO,
  LINE_TYPE,
  MARGIN,
  ROTATION,
  SCALE_BAND,
  SCALE_LINEAR,
  SCALE_TIME,
  SECONDARY_THEME,
  SIZE,
  THEME,
  TICKS,
} from '../utils/constants';

const BarLine = ({
  aspectRatio = ASPECT_RATIO,
  data: chartData,
  grid,
  height: svgHeight = undefined,
  isHorizontal,
  lineSeries = [],
  lineType = LINE_TYPE,
  lineTypeOption = null,
  margin = MARGIN,
  onClick = identity,
  onMouseEnter = identity,
  onMouseLeave = identity,
  responsive = false,
  secondaryTheme = SECONDARY_THEME,
  stackedSeries = [],
  theme = THEME,
  ticks = TICKS,
  width: svgWidth = undefined,
  xAxisLabel,
  xAxisLabelRotation,
  xAxisLabelRotationValue = ROTATION,
  yAxisLabel,
}) => {
  const svgRef = useRef();
  const [id] = useState(getId('bar-line'));
  const [{ width, height, isSizeSet }, setSize] = useState(SIZE);
  const [isDates, data] = useMemo(() => setupData(chartData), chartData);
  const stack = useMemo(
    () => buildStack(stackedSeries)(toStackedForm(data)),
    data
  );

  const xScale = getXScale(
    isDates ? SCALE_TIME : SCALE_BAND,
    data,
    width,
    true
  );
  const yScale = getYScale(
    SCALE_LINEAR,
    getMax(getStackedMax(data, stackedSeries)),
    height
  );

  const line = d3Line()
    .curve(setLineType(lineType, lineTypeOption))
    .x(({ name }) =>
      isDates ? xScale(name) : xScale(name) + xScale.bandwidth() / 2
    )
    .y(({ value }) => yScale(value));

  const lineData = getLineDataForSeries(lineSeries, data);

  const handleSize = () => {
    const offsetWidth = svgRef.current.parentElement.offsetWidth;
    if ((svgWidth || svgHeight) && !isSizeSet) {
      setSize({
        ...getSize(svgWidth, svgHeight, margin, aspectRatio),
        isSizeSet: true,
      });
    } else if (offsetWidth !== svgWidth - (margin.left + margin.right)) {
      setSize({
        ...getSize(offsetWidth, undefined, margin, aspectRatio),
        isSizeSet: true,
      });
    }
  };

  const handleResize = debounce(handleSize)();

  useEffect(() => {
    handleSize();
    responsive && window.addEventListener('resize', handleResize);

    return () => {
      responsive && window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <SVG
      identifier={id}
      size={{
        width: svgWidth || width + margin.left + margin.right,
        height: svgHeight || height + margin.top + margin.bottom,
      }}
      ref={svgRef}
    >
      <g
        className="silky-charts-container"
        transform={`translate(${margin.left}, ${margin.top})`}
      >
        {grid && (
          <Grid
            ref={node =>
              d3Select(node).call(
                drawGrid(isHorizontal, xScale, height, yScale, width, ticks)
              )
            }
          />
        )}

        {xAxisLabel && (
          <Label axis="x" margin={margin} width={width} height={height}>
            {xAxisLabel}
          </Label>
        )}

        {yAxisLabel && (
          <Label axis="y" margin={margin} width={width} height={height}>
            {yAxisLabel}
          </Label>
        )}

        <StackedBarDatum
          data={data}
          series={stack}
          isDates={isDates}
          theme={theme}
          x={xScale}
          y={yScale}
          width={width}
          height={height}
          onClick={onClick}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
        />

        <Axis
          axis="x"
          position={{ x: 0, y: height }}
          ref={node => {
            d3Select(node).call(d3AxisBottom(xScale));
            isDates && extendXPath(id, width);
            xAxisLabelRotation && rotateXLabels(id, xAxisLabelRotationValue);
          }}
        />
        <Axis
          axis="y"
          ref={node => d3Select(node).call(d3AxisLeft(yScale).ticks(ticks))}
        />

        {lineData.map((datum, idx) => (
          <g className={`${head(datum)['series']}-layer`} key={idx}>
            <LineDatum
              chart="bar-line"
              data={datum}
              isDates={isDates}
              color={palette.themes[secondaryTheme].base[idx]}
              d={line(datum)}
              xScale={xScale}
              yScale={yScale}
              onClick={onClick}
              onMouseEnter={onMouseEnter}
              onMouseLeave={onMouseLeave}
            />
          </g>
        ))}
      </g>
    </SVG>
  );
};

export default BarLine;
