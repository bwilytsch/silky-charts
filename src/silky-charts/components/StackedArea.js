import React, { useEffect, useMemo, useRef, useState } from 'react';
import identity from 'ramda/src/identity';
import { axisBottom as d3AxisBottom, axisLeft as d3AxisLeft } from 'd3-axis';
import { select as d3Select } from 'd3-selection';
import { area as d3Area } from 'd3-shape';
import { Axis, Grid, Label, Path, SVG } from './styled';
import {
  appendStackedValues,
  buildStack,
  bySeries,
  classify,
  debounce,
  drawGrid,
  getId,
  getMax,
  getStackedMax,
  getSeries,
  getSize,
  getXScale,
  getYScale,
  palette,
  rotateXLabels,
  setLineType,
  setupData,
  toStackedForm,
} from '../utils';
import {
  ASPECT_RATIO,
  LINE_TYPE,
  MARGIN,
  ROTATION,
  SCALE_BAND,
  SCALE_LINEAR,
  SCALE_TIME,
  SIZE,
  THEME,
  TICKS,
} from '../utils/constants';

const StackedArea = ({
  aspectRatio = ASPECT_RATIO,
  data: chartData,
  grid,
  height: svgHeight = undefined,
  isHorizontal,
  lineSeries = [],
  lineType = LINE_TYPE,
  lineTypeOption = null,
  margin = MARGIN,
  onClick = identity,
  onMouseEnter = identity,
  onMouseLeave = identity,
  responsive = false,
  theme = THEME,
  ticks = TICKS,
  width: svgWidth = undefined,
  xAxisLabel,
  xAxisLabelRotation,
  xAxisLabelRotationValue = ROTATION,
  yAxisLabel,
}) => {
  const svgRef = useRef();
  const [id] = useState(getId('stacked-area'));
  const [{ width, height, isSizeSet }, setSize] = useState(SIZE);
  let [isDates, data] = useMemo(() => setupData(chartData), chartData);
  data = useMemo(
    () =>
      appendStackedValues(
        buildStack(getSeries(data))(toStackedForm(data)),
        data
      ),
    data
  );

  const xScale = getXScale(isDates ? SCALE_TIME : SCALE_BAND, data, width);
  const yScale = getYScale(SCALE_LINEAR, getMax(getStackedMax(data)), height);

  const area = d3Area()
    .curve(setLineType(lineType, lineTypeOption))
    .x(({ name }) => xScale(name))
    .y0(({ stackedValues }) => yScale(stackedValues[0]))
    .y1(({ stackedValues }) => yScale(stackedValues[1]));

  const handleSize = () => {
    const offsetWidth = svgRef.current.parentElement.offsetWidth;
    if ((svgWidth || svgHeight) && !isSizeSet) {
      setSize({
        ...getSize(svgWidth, svgHeight, margin, aspectRatio),
        isSizeSet: true,
      });
    } else if (offsetWidth !== svgWidth - (margin.left + margin.right)) {
      setSize({
        ...getSize(offsetWidth, undefined, margin, aspectRatio),
        isSizeSet: true,
      });
    }
  };

  const handleResize = debounce(handleSize)();

  useEffect(() => {
    handleSize();
    responsive && window.addEventListener('resize', handleResize);

    return () => {
      responsive && window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <SVG
      identifier={id}
      size={{
        width: svgWidth || width + margin.left + margin.right,
        height: svgHeight || height + margin.top + margin.bottom,
      }}
      ref={svgRef}
    >
      <g
        className="silky-charts-container"
        transform={`translate(${margin.left}, ${margin.top})`}
      >
        {grid && (
          <Grid
            ref={node =>
              d3Select(node).call(
                drawGrid(isHorizontal, xScale, height, yScale, width, ticks)
              )
            }
          />
        )}

        {xAxisLabel && (
          <Label axis="x" margin={margin} width={width} height={height}>
            {xAxisLabel}
          </Label>
        )}

        {yAxisLabel && (
          <Label axis="y" margin={margin} width={width} height={height}>
            {yAxisLabel}
          </Label>
        )}

        {bySeries(data).map(([series, datum], idx) => (
          <g className={`${classify(series)}-layer`} key={idx}>
            <Path
              chart="stacked-area"
              fillColor={palette.themes[theme].base[idx]}
              d={area(datum)}
              strokeWidth={0}
              onClick={onClick}
              onMouseEnter={onMouseEnter}
              onMouseLeave={onMouseLeave}
            />
          </g>
        ))}

        <Axis
          axis="x"
          position={{ x: 0, y: height }}
          ref={node => {
            d3Select(node).call(d3AxisBottom(xScale));
            xAxisLabelRotation && rotateXLabels(id, xAxisLabelRotationValue);
          }}
        />
        <Axis
          axis="y"
          ref={node => d3Select(node).call(d3AxisLeft(yScale).ticks(ticks))}
        />
      </g>
    </SVG>
  );
};

export default StackedArea;
