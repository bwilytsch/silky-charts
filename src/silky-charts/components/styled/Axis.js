import styled from 'styled-components';

const Axis = styled.g.attrs(({ axis, translate }) => ({
  className: `axis-${axis}`,
  transform: translate && `translate(${translate.x}, ${translate.y})`,
}))``;

export default Axis;
