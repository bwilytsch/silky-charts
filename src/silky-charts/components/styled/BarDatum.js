import React, { Fragment, useState } from 'react';
import { createPortal } from 'react-dom';
import { Rect, Tooltip, TooltipItem } from './';

const BarDatum = ({
  color,
  datum,
  x,
  y,
  width,
  height,
  onClick,
  onMouseEnter,
  onMouseLeave,
}) => {
  const [tooltip, setTooltip] = useState({
    pageX: null,
    pageY: null,
    show: false,
  });

  return (
    <Fragment>
      <Rect
        chart="bar"
        position={{ x, y }}
        size={{ width, height }}
        fillColor={color}
        onClick={onClick}
        onMouseEnter={event => {
          setTooltip(state => ({ ...state, show: true }));
          onMouseEnter(event);
        }}
        onMouseLeave={event => {
          setTooltip(state => ({ ...state, show: false }));
          onMouseLeave(event);
        }}
        onMouseMove={event => {
          event.persist();
          const { pageX, pageY } = event;
          setTooltip(state => ({ ...state, pageX, pageY }));
        }}
      />
      {tooltip.show &&
        createPortal(
          <Tooltip pageX={tooltip.pageX} pageY={tooltip.pageY}>
            <TooltipItem color={color} {...datum} />
          </Tooltip>,
          document.body
        )}
    </Fragment>
  );
};

export default BarDatum;
