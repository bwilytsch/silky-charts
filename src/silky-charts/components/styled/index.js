export { default as Axis } from './Axis';
export { default as BarDatum } from './BarDatum';
export { default as Circle } from './Circle';
export { default as Grid } from './Grid';
export { default as Label } from './Label';
export { default as LineDatum } from './LineDatum';
export { default as Path } from './Path';
export { default as Rect } from './Rect';
export { default as StackedBarDatum } from './StackedBarDatum';
export { default as SVG } from './SVG';
export { default as Tooltip } from './Tooltip';
export { default as TooltipItem } from './TooltipItem';
