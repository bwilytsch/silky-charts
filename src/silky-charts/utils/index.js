export { default as allDate } from './allDate';
export { default as drawGrid } from './drawGrid';
export { default as extendXPath } from './extendXPath';
export { default as getDivergence } from './getDivergence';
export { default as getXScale } from './getXScale';
export { default as getYScale } from './getYScale';
export { default as rotateXLabels } from './rotateXLabels';
export { default as themes } from './themes';
