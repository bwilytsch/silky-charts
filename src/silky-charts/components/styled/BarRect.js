import React, { Fragment } from 'react';
import styled from 'styled-components';
import { bar, black, white } from '../../utils/themes';

const Rect = styled.rect.attrs(({ position: { x, y } }) => ({
  x,
  y,
  className: 'bar',
}))`
  fill: ${({ theme }) => bar[theme].bg};
  height: ${({ size }) => size.height}px;
  width: ${({ size }) => size.width}px;

  &:hover {
    fill: ${({ theme }) => bar[theme].fg};
  }
`;

const Text = styled.text.attrs(
  ({ position: { x, y }, size: { width, height } }) => ({
    className: 'value',
    textAnchor: 'middle',
    transform: `translate(${x + width / 2}, ${y + (height < 40 ? -18 : 30)})`,
  })
)`
  fill: ${({ size: { height } }) => (height < 40 ? black : white)};
  pointer-events: none;
`;

const BarRect = ({
  datum: { name, value },
  position,
  x,
  y,
  width,
  height,
  onClick,
  onMouseEnter,
  onMouseLeave,
  showValue,
  theme,
}) => (
  <Fragment>
    <Rect
      position={{ x, y }}
      size={{ width, height }}
      theme={theme}
      onClick={onClick}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {value}
    </Rect>
    {showValue && (
      <Text position={{ x, y }} size={{ width, height }}>
        {value}
      </Text>
    )}
  </Fragment>
);

export default BarRect;
