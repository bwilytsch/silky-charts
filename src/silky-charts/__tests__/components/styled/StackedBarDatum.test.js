import React from 'react';
import { StackedBarDatum } from '../../../components/styled';
import { create } from 'react-test-renderer';
import { buildStack } from '../../../utils';
import 'jest-styled-components';

const stack = buildStack(['bar', 'baz']);

const series = stack([
  { name: 'foo', foo: 0, bar: 1, baz: 2, qux: 3 },
  { name: 'baz', foo: 4, bar: 5, baz: 6, qux: 7 },
]);

const seriesWithDate = stack([
  { name: '2019-01-01T05:00:00.000Z', foo: 0, bar: 1, baz: 2, qux: 3 },
  { name: '2019-01-02T05:00:00.000Z', foo: 4, bar: 5, baz: 6, qux: 7 },
]);

const xScale = jest.fn();
xScale.bandwidth = jest.fn();

const props = {
  theme: 'red',
  x: xScale,
  y: jest.fn(),
  width: 10,
  height: 10,
};

test('render correctly', () => {
  const tree = create(<StackedBarDatum series={series} {...props} />).toJSON();

  expect(tree).toMatchSnapshot();
});

test('render correctly with date', () => {
  const tree = create(
    <StackedBarDatum
      series={seriesWithDate}
      data={[{ name: '2019-01-01', value: 0 }]}
      {...props}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
