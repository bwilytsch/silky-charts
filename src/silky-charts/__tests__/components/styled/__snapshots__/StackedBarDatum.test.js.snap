// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
Array [
  .c0 {
  fill: rgb(127,29,24);
  height: NaNpx;
  width: px;
}

.c0:hover {
  fill: hsl(2.8999999999999773,68.2%,23.7%);
}

<g
    className="bar-layer"
  >
    <rect
      className="bar c0"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      size={
        Object {
          "height": NaN,
          "width": undefined,
        }
      }
    />
    <rect
      className="bar c0"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      size={
        Object {
          "height": NaN,
          "width": undefined,
        }
      }
    />
  </g>,
  .c0 {
  fill: rgb(255,131,124);
  height: NaNpx;
  width: px;
}

.c0:hover {
  fill: hsl(3.1999999999999886,100%,59.5%);
}

<g
    className="baz-layer"
  >
    <rect
      className="bar c0"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      size={
        Object {
          "height": NaN,
          "width": undefined,
        }
      }
    />
    <rect
      className="bar c0"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      size={
        Object {
          "height": NaN,
          "width": undefined,
        }
      }
    />
  </g>,
]
`;

exports[`render correctly with date 1`] = `
Array [
  .c0 {
  fill: rgb(127,29,24);
  height: NaNpx;
  width: px;
}

.c0:hover {
  fill: hsl(2.8999999999999773,68.2%,23.7%);
}

<g
    className="bar-layer"
  >
    <rect
      className="bar c0"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      size={
        Object {
          "height": NaN,
          "width": undefined,
        }
      }
    />
    <rect
      className="bar c0"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      size={
        Object {
          "height": NaN,
          "width": undefined,
        }
      }
    />
  </g>,
  .c0 {
  fill: rgb(255,131,124);
  height: NaNpx;
  width: px;
}

.c0:hover {
  fill: hsl(3.1999999999999886,100%,59.5%);
}

<g
    className="baz-layer"
  >
    <rect
      className="bar c0"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      size={
        Object {
          "height": NaN,
          "width": undefined,
        }
      }
    />
    <rect
      className="bar c0"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      size={
        Object {
          "height": NaN,
          "width": undefined,
        }
      }
    />
  </g>,
]
`;
